File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.py
import math

def solution(arrayA, arrayB):
    answer = 0
    
    lengthA = len(arrayA)
    lengthB = len(arrayB)
    gcdA = arrayA[0]
    gcdB = arrayB[0]
    
    # array길이 2 이상
    if lengthA >= 2:
        for i in range(1, lengthA):
            gcdA = math.gcd(gcdA, arrayA[i])
    if lengthB >= 2:
        for i in range(1, lengthB):
            gcdB = math.gcd(gcdB, arrayB[i])
    
    # a가 될 수 있는지
    aA = True
    aB = True
    
    if gcdA == 1:
        aA = False
    else:
        for b in arrayB:
            if b % gcdA == 0:
                aA = False
    if gcdB == 1:
        aB = False
    else:
        for a in arrayA:
            if a % gcdB == 0:
                aB = False
    
    if not aA and not aB:
        return 0
    elif aA and not aB:
        return gcdA
    elif not aA and aB:
        return gcdB
    else:
        return max(gcdA, gcdB)
    
    return answer

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
import java.util.*;

class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        List<Integer> divisorsA = getAllCd(arrayA);
        List<Integer> divisorsB = getAllCd(arrayB);
        
        int answer = Math.max(
            findAnswer(arrayB, divisorsA), 
            findAnswer(arrayA, divisorsB)
        );
        
        return answer;
    }
    
    private List<Integer> getAllCd(int[] array) {
        int value = array[0];
        for (int i = 1; i < array.length; i++) {
            value = gcd(value, array[i]);
        }
        
        List<Integer> divisors = new ArrayList<>();
        for (int i = 1; i * i <= value; i++) {
            if (value % i == 0) {
                divisors.add(i);
                if (value / i != i)
                    divisors.add(value / i);
            }
        }
        Collections.sort(divisors, Collections.reverseOrder());
        return divisors;
    }
    
    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
    
    private int findAnswer(int[] array, List<Integer> divisors) {
        int answer = 0;
        for (int divisor : divisors) {
            boolean possible = true;
            for (int num : array) {
                if (num % divisor == 0) {
                    possible = false;
                    break;
                }
            }
            if (possible) {
                answer = divisor;
                break;
            }
        }
        return answer;
    }
}